setup.py

```diff
@@ -668,11 +668,50 @@ def is_mx_mkldnn():
                 return os.environ.get('MXNET_USE_MKLDNN', '0') == '1'
 
 
+def is_mx_cuda():
+    try:
+        from mxnet import runtime
+        features = runtime.Features()
+        return features.is_enabled('CUDA')
+    except Exception:
+        if 'linux' in sys.platform:
+            try:
+                import mxnet as mx
+                mx_libs = mx.libinfo.find_lib_path()
+                for mx_lib in mx_libs:
+                    output = subprocess.check_output(['readelf', '-d', mx_lib])
+                    if 'cuda' in str(output):
+                        return True
+                return False
+            except Exception:
+                return False
+    return False
+
+
 def build_mx_extension(build_ext, options):
     check_mx_version()
     mx_compile_flags, mx_link_flags = get_mx_flags(
         build_ext, options['COMPILE_FLAGS'])
 
+    mx_have_cuda = is_mx_cuda()
+    macro_have_cuda = check_macro(options['MACROS'], 'HAVE_CUDA')
+    if not mx_have_cuda and macro_have_cuda:
+        raise DistutilsPlatformError(
+            'Horovod build with GPU support was requested, but this MXNet '
+            'installation does not support CUDA.')
+
+    # Update HAVE_CUDA to mean that MXNet supports CUDA. Internally, we will be checking
+    # HOROVOD_GPU_(ALLREDUCE|ALLGATHER|BROADCAST) to decide whether we should use GPU
+    # version or transfer tensors to CPU memory for those operations.
+    if mx_have_cuda and not macro_have_cuda:
+        cuda_include_dirs, cuda_lib_dirs = get_cuda_dirs(build_ext, options['COMPILE_FLAGS'])
+        options['MACROS'] += [('HAVE_CUDA', '1')]
+        options['INCLUDES'] += cuda_include_dirs
```

Comment:
Do we need anything other than MACRO update here?  AFAIK other frameworks don't have this logic.  You can also use `updated_macros = set_macro(options['MACROS'], 'HAVE_CUDA', str(int(mx_have_cuda)))` for safety.

Reply:
I did what you mentioned in my very first [commit](input) in this PR. However, it cannot compile if I do not set HOROVOD_GPU_ALLREDUCE/BROADCAST/ALLGATHER.

I do not know how other frameworks worked. But for MXNet, I think we need to call `get_cuda_dirs` here when HOROVOD_GPU_ALLREDUCE/BROADCAST/ALLGATHER is not set. Right now, it is [called](https://github.com/horovod/horovod/blob/master/setup.py#L504) in `get_comm_options`. If those options are not set, it will be skipped here. Therefore, if HAVE_CUDA is not set in MACROS and we detect CUDA is supported in the MXNet build, we need to set the include/lib/macro etc. for CUDA.

Reply:
What's the error message if you just set the macro?

Reply:
This was the error:
```
    In file included from horovod/common/operations.cc:45:0:
    horovod/common/ops/cuda_operations.h:23:26: fatal error: cuda_runtime.h: No such file or directory
    compilation terminated.
```

Reply:
Oh, I see. Fair enough :-)